### ISSUE TITLE
Generalised linear models in Stan

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3824]
Generalised linear models in Stan<br>
<br>
http://bit.ly/2V1dmMT<br>
<br>
About 0 Minutes<br>
<br>
<div><div>

<section>
<div>
<div>
<div>
<div>
<p>By <span>Gergana and Maxwell Farrell</span> on <span>April 30, 2018</span></p>
<hr>
<div>
<center>
<img alt="Img" src="https://ourcodingclub.github.io/img/tutheader_stan2.png">
</center>
</div>
<h3>Tutorial Aims:</h3>
<h4><a href="#model"> 1. Learn about generalised models in <code>Stan</code></a></h4>
<h4><a href="#rstanarm"> 2. Use the <code>rstanarm</code> package to run a <code>Poisson</code> model</a></h4>
<h4><a href="#assess"> 3. Assess model convergence </a></h4>
<h4><a href="#priors"> 4. Check priors in <code>rstanarm</code> </a></h4>
<h4><a href="#negbin"> 6. Run a model with a negative binomial distribution </a></h4>
<h4><a href="#brms"> 7. Compare <code>rstanarm</code> and <code>brms</code></a></h4>
<h3>All the files you need to complete this tutorial can be downloaded from <a href="https://github.com/ourcodingclub/CC-Stan-2" target="_blank">this repository</a>. Click on <code>Clone/Download/Download ZIP</code> and unzip the folder, or clone the repository to your own GitHub account.</h3>
<h2>Introduction</h2>
<p><strong>Finding answers to our research questions often requires statistical models. Designing models, choosing what variables to include, which data distribution to use are all worth thinking about carefully. In this tutorial, we will continue exploring different model structures in search of the best way to find the answers to our research questions. We will build on the Coding Club tutorials on <a href="https://ourcodingclub.github.io/2018/04/06/model-design.html" target="_blank"> how to design a model</a>, and on<a href="https://ourcodingclub.github.io/2018/01/22/mcmcglmm.html" target="_blank">Bayesian Modelling in <code>MCMCglmm</code></a> for key background information on model design and Bayesian statistics.</strong></p>
<p><strong>Statistical models can be fit in a variety of packages in <code>R</code> or other statistical languages. But sometimes the perfect model that you can design conceptually is very hard or impossible to implement in a package or programme that restricts the distributions and complexity that you can use. This is when you may want to move to a statistical programming language such as <a href="http://mc-stan.org/" target="_blank"><code>Stan</code></a>. For an introduction to <code>Stan</code>, you can check out our <a href="https://ourcodingclub.github.io/2018/04/17/stan-intro.html" target="_blank">intro tutorial here</a>.</strong></p>
<p>In this tutorial, we will learn about two packages, <code>rstanarm</code> and <code>brms</code> which allow us to fit <code>Stan</code> models using syntax similar to packages like <code>lme4</code>, <code>nlme</code> and <code>MCMCglmm</code>. We will use these packages to fit models that test how species richness has changed over time near <a href="http://arc-lter.ecosystems.mbl.edu/terrestrial-data" target="_blank">Toolik Lake Field Station</a>.</p>
<h3>Research question: How has plant species richness changed over time at Toolik Lake?</h3>
<h3>Hypothesis: Plant species richness has increased over time at Toolik Lake.</h3>
<p>We can start with loading the libraries we will need and the data we will use. If you don’t have some of these packages installed, you can install them using <code>install.packages("package-name")</code>.</p>
<div><div><pre><code><span># Load libraries ----</span><span>
</span><span>library</span><span>(</span><span>rstanarm</span><span>)</span><span>
</span><span>library</span><span>(</span><span>brms</span><span>)</span><span> </span><span># for models</span><span>
</span><span>library</span><span>(</span><span>bayesplot</span><span>)</span><span>
</span><span>library</span><span>(</span><span>ggplot2</span><span>)</span><span>
</span><span>library</span><span>(</span><span>dplyr</span><span>)</span><span>
</span><span>library</span><span>(</span><span>tidybayes</span><span>)</span><span>
</span><span>library</span><span>(</span><span>modelr</span><span>)</span><span>
</span><span># Load data ----</span><span>
</span><span># Remember to set your working directory to the folder</span><span>
</span><span># where you saved the workshop files</span><span>
</span><span>toolik_richness</span><span> </span><span>&lt;-</span><span> </span><span>read.csv</span><span>(</span><span>"toolik_richness.csv"</span><span>)</span><span>
</span><span># Inspect data</span><span>
</span><span>head</span><span>(</span><span>toolik_richness</span><span>)</span><span>
</span></code></pre></div></div>
<p>To check out what class of data we are dealing with we can use the str() function.</p>
<p><code>Site</code> and <code>Species</code> are strings (letters) and categorical data (factors) - they are names. <code>Year</code>, <code>Cover</code>, <code>Mean.Temp</code> and <code>SD.Temp</code> are numeric and continuous data - they are numbers. <code>Cover</code> shows the relative cover (out of 1) for different plant species, <code>Mean.Temp</code> is the mean annual temperature at Toolik Lake Station and <code>SD.Temp</code> is the standard deviation of the mean annual temperature. Then we have <code>Treatment</code>, another categorical variable that refers to different chemical treatments, e.g. some plots received extra nitrogen, others extra phosphorus. Finally, we have <code>Block</code> and <code>Plot</code>, which give more detailed information about where the measurements were taken.</p>
<p>The plot numbers are currently coded as numbers - 1, 2,…8 and they are a numerical variable. We should make them a categorical variable, since just like <code>Site</code> and <code>Block</code>, the numbers represent the different categories, not actual count data.</p>
<div><div><pre><code><span>toolik_richness</span><span>$</span><span>Plot</span><span> </span><span>&lt;-</span><span> </span><span>as.factor</span><span>(</span><span>as.character</span><span>(</span><span>toolik_richness</span><span>$</span><span>Plot</span><span>))</span><span>
</span></code></pre></div></div>
<p>Now, let’s think about the distribution of the data, specifically our response variable, species richness (<code>Richness</code>).</p>
<div><div><pre><code><span>(</span><span>hist</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>(</span><span>toolik_richness</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>Richness</span><span>))</span><span> </span><span>+</span><span>
</span><span>geom_histogram</span><span>()</span><span> </span><span>+</span><span>
</span><span>theme_classic</span><span>())</span><span>
</span></code></pre></div></div>
<p>Note that putting your entire ggplot code in brackets () creates the graph and then shows it in the plot viewer. If you don’t have the brackets, you’ve only created the object, but haven’t visualized it. You would then have to call the object such that it will be displayed by just typing <code>hist</code> after you’ve created the “hist” object.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/richness_hist.png"> </center>
<p><strong>We are working with integers (similar to count data, species only come in whole numbers) and the data are right-skewed, that is because we have a few data points on the extreme right end, they are pulling the mean and median or our data to the right. For ecological data, this is quite common - across all of the sampling plots, we expect that they won’t all have lots of different species within them.</strong></p>
<p><strong>This means that a <code>Poisson</code> distribution might be suitable for our model.</strong></p>
<p>One advantage to fitting models in <code>Stan</code> is that it’s easy to fully take advantage of your computing power. In <code>Stan</code>, we usually run two or more <strong>chains</strong> - different iterations of our model which we can then compare - if they are massively different, something is not right. If your computer has two or more cores, you can split the chains, i.e., run a chain on each core, saving you some time, as the code will then finish running quicker. This means that you’ll be running the code in parallel.</p>
<p>To enable parallel computing, you can run this line of code and then later on in the model code, you can specify how many cores you want to use.</p>
<div><div><pre><code><span>options</span><span>(</span><span>mc.cores</span><span> </span><span>=</span><span> </span><span>parallel</span><span>::</span><span>detectCores</span><span>())</span><span>
</span></code></pre></div></div>
<p><strong>Now we are all set up for our first model. Remember that the data have a hierarchical structure - species richness is measured in plots, which fall within blocks that are then part of different sites. To derive inferences about changes species richness through time, our models should take this complexity of the data structure into account.</strong></p>
<p>One disadvantage to <code>Stan</code> models is that the code can take a while to finish running, even if we use all of our computer cores, it can still be hours before you can see a summary of your model. Of course, one should use the most suitable type of model for a research question, regardless of how long the model takes to run (within reason…), but in our case, for teaching purposes and so that you can see some model outputs today, we will focus on species richness change within just one of the plots. In that case, the model does not need to include random effects, because on the plot level, there is no replication.</p>
<p>Just so that you know what the syntax looks like and if you have time to wait for this code to finish running, this is how the model for species richness change around Toolik Lake in general would look like:</p>
<div><div><pre><code><span># Note - this code could take hours to run!</span><span>
</span><span># We are running the model using the default weakly informative priors.</span><span>
</span><span># More about priors coming later in the tutorial!</span><span>
</span><span># stan_lm &lt;- stan_glmer(Richness ~ I(Year-2007) + (1|Site/Block/Plot),</span><span>
</span><span># data = toolik_richness, family = poisson,</span><span>
</span><span># chains = 4, cores = 4)</span><span>
</span><span># Assess converge by looking at the trace plots</span><span>
</span><span># plot(stan_lm, plotfun = "trace")</span><span>
</span><span># Explore the summary output</span><span>
</span><span># summary(stan_lm)</span><span>
</span></code></pre></div></div>
<p><strong>For teaching purposes only, we will proceed with a model without random effects - that way you can see the outputs of the models as you are going through the tutorial. Note that we do not advise doing this when you are analysing your data for real - of course a model that takes into account the hierarchical structure of the data would be better.</strong></p>
<h3>Advantages and disadvantages of using <code>Stan</code>
</h3>
<p><code>Stan</code> models can take a long time to compile. One of their key advantage is that you have a lot of freedom to specify the priors (your prior knowledge and expectations of how a given parameter will behave) for your model parameters and you can really account for the complex structure your data might have. There is a time cost to this, as we’ve seen above. One way to approach this is to first make sure your code works on a small chunk of your data, and only afterwards, you can start running the code on the full data (and do other things while you wait for the models to compile).</p>
<p>Now we can run our simplified model. First, let’s check how many years of data we have:</p>
<div><div><pre><code><span>unique</span><span>(</span><span>toolik_richness</span><span>$</span><span>Year</span><span>)</span><span>
</span></code></pre></div></div>
<p>There are four years of data from 2008 to 2012. For modelling purposes, it helps to transform the year variable into a continuous variable where the first year is year one, i.e., 2008 is 1, 2009 is 2. That way, the model won’t have to estimate richness in the year 500, the year 501, etc., it will start straight from our first monitoring year.</p>
<div><div><pre><code><span># Note how now we are using stan_glm because</span><span>
</span><span># there are no random effects</span><span>
</span><span>stan_glm1</span><span> </span><span>&lt;-</span><span> </span><span>stan_glm</span><span>(</span><span>Richness</span><span> </span><span>~</span><span> </span><span>I</span><span>(</span><span>Year</span><span>-2007</span><span>),</span><span>
</span><span>data</span><span> </span><span>=</span><span> </span><span>toolik_richness</span><span>,</span><span> </span><span>family</span><span> </span><span>=</span><span> </span><span>poisson</span><span>,</span><span>
</span><span>chains</span><span> </span><span>=</span><span> </span><span>4</span><span>,</span><span> </span><span>cores</span><span> </span><span>=</span><span> </span><span>4</span><span>)</span><span>
</span></code></pre></div></div>
<p>If you find this code still takes a long time, you can change the <code>chains</code> argument to only two chains, but note that it’s better to run models with more than two chains - then you have more room for comparison. If one or more of the four chains is behaving really differently from the rest, then the model might not have converged. What is model convergence? In brief, if a model hasn’t converged, you can’t trust the estimates it gives you. You can find more details in <a href="https://ourcodingclub.github.io/2018/04/06/model-design.html" target="_blank"> the model design tutorial here</a>.</p>
<h2>Assessing model convergence</h2>
<p><strong>One way to assess model convergence is by visually examining the trace plots. They should be fuzzy with no big gaps, breaks or gigantic spikes.</strong></p>
<div><div><pre><code><span>plot</span><span>(</span><span>stan_glm1</span><span>,</span><span> </span><span>plotfun</span><span> </span><span>=</span><span> </span><span>"trace"</span><span>)</span><span>
</span></code></pre></div></div>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/stan2_traces.png"> </center>
<p>Here, the trace plots look fine.</p>
<p>Next we can look at the summary output.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/stan2_summary.png"> </center>
<p><strong>We can see that the effective sample size is alright (there is no hard cut threshold, but more than around 1000 is usually a good sign), another diagnostic metric, the <code>Rhat</code> value is also indicating convergence (an <code>Rhat</code> of 1 is a good sign, more than 1 could indicate trouble).</strong></p>
<h2>Posterior predictive checks</h2>
<p>The pre-compiled models in <code>rstanarm</code> already include a <code>y_rep</code> variable (our model predictions) in the generated quantities block (your posterior distributions). We can use the <code>pp_check</code> function from the <code>bayesplot</code> package to see how the model predictions compare to the raw data, i.e., is the model behaving as we expect it to be?</p>
<div><div><pre><code><span>pp_check</span><span>(</span><span>stan_glm1</span><span>,</span><span> </span><span>plotfun</span><span> </span><span>=</span><span> </span><span>"stat"</span><span>,</span><span> </span><span>stat</span><span> </span><span>=</span><span> </span><span>"mean"</span><span>)</span><span>
</span><span>pp_check</span><span>(</span><span>stan_glm1</span><span>,</span><span> </span><span>plotfun</span><span> </span><span>=</span><span> </span><span>"dens_overlay"</span><span>)</span><span>
</span></code></pre></div></div>
<center>
<img alt="Img" src="https://ourcodingclub.github.io/img/stan2_hist.png"> <img alt="Img" src="https://ourcodingclub.github.io/img/stan2_density.png"> </center>
<p><strong>What do you think? How does the density distribution of the model predictions compare with that of the raw data?</strong></p>
<p>From the histogram, we can see that the posterior mean (the dark blue line) aligns well with the raw data. From the density plot, we can see that the model is underpredicting the low species richness value. This is a common problem when working with left-skewed or zero/low number-inflated data. Overall though, the model predictions follow the underlying data, so in summary, we can say that this model fit is acceptable. Note that these decisions can vary based on your question, data and even your statistical philosophy.</p>
<h2>Model diagnostics using <code>shinystan</code>
</h2>
<p>For an interactive exploration of <code>Stan</code> models, you can use the <code>shinystan</code> app. The app is compatible with <code>Stan</code> models generated using the <code>rstan</code>, <code>rstanarm</code> and <code>brms</code> packages, so regardless of how you chooose to run your <code>Stan</code> models, you can still use <code>shinystan</code> to assess whether or not they’ve converged and are behaving properly.</p>
<p>We can launch the app using the code below. That will open a window in our internet browser, where we can further explore our model.</p>
<div><div><pre><code><span>launch_shinystan</span><span>(</span><span>stan_glm1</span><span>)</span><span>
</span></code></pre></div></div>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/shinystan.png"> </center>
<p>Have a go at exploring the various options - you’ll probably spot some of the plots we’ve already made in <code>R</code>, but there are many others as well. Here, for example, if there were any divergent chains (i.e. a chain that is behaving in a very weird unexpected way), we would have seen red points. In our case, we don’t have any divergent chains.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/shinystan2.png"> </center>
<h3>Back to our research question</h3>
<p><strong>How has species richness changed over those four years near Toolik lake?</strong></p>
<p>To get the answer, we can plot the model predictions for our model as well as the raw data points.</p>
<div><div><pre><code><span>(</span><span>model_fit</span><span> </span><span>&lt;-</span><span> </span><span>toolik_richness</span><span> </span><span>%&gt;%</span><span>
</span><span>data_grid</span><span>(</span><span>Year</span><span> </span><span>=</span><span> </span><span>seq_range</span><span>(</span><span>Year</span><span>,</span><span> </span><span>n</span><span> </span><span>=</span><span> </span><span>101</span><span>))</span><span> </span><span>%&gt;%</span><span>
</span><span>add_predicted_draws</span><span>(</span><span>stan_glm1</span><span>)</span><span> </span><span>%&gt;%</span><span>
</span><span>ggplot</span><span>(</span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>Year</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>Richness</span><span>))</span><span> </span><span>+</span><span>
</span><span>stat_lineribbon</span><span>(</span><span>aes</span><span>(</span><span>y</span><span> </span><span>=</span><span> </span><span>.prediction</span><span>),</span><span> </span><span>.width</span><span> </span><span>=</span><span> </span><span>c</span><span>(</span><span>.95</span><span>,</span><span> </span><span>.80</span><span>,</span><span> </span><span>.50</span><span>),</span><span>
</span><span>alpha</span><span> </span><span>=</span><span> </span><span>1</span><span>/</span><span>2</span><span>,</span><span> </span><span>colour</span><span> </span><span>=</span><span> </span><span>"black"</span><span>)</span><span> </span><span>+</span><span>
</span><span>geom_point</span><span>(</span><span>data</span><span> </span><span>=</span><span> </span><span>toolik_richness</span><span>,</span><span> </span><span>colour</span><span> </span><span>=</span><span> </span><span>"darkseagreen4"</span><span>,</span><span> </span><span>size</span><span> </span><span>=</span><span> </span><span>3</span><span>)</span><span> </span><span>+</span><span>
</span><span>scale_fill_brewer</span><span>(</span><span>palette</span><span> </span><span>=</span><span> </span><span>"Greys"</span><span>))</span><span>
</span></code></pre></div></div>
<p><strong>Looks like species richness is decreasing, but also, four years of data is not enough to really test temporal dynamics!</strong></p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/stan_pred.png"> </center>
<h2>Priors</h2>
<p>Packages like <code>rstanarm</code> and <code>brms</code> allow us to fit <code>Stan</code> models using simple and quick code syntax. One danger though is that along the way, we might forget to think about our priors! In the code above, we have not specified any priors. In that case, the model uses the default <code>rstanarm</code> priors.</p>
<p>We should check what those are whether they match our expectations of the data.</p>
<div><div><pre><code>Priors for model 'stan_glm1'
------
Intercept (after predictors centered)
 ~ normal(location = 0, scale = 10)

Coefficients
 ~ normal(location = 0, scale = 2.5)
     **adjusted scale = 1.58
</code></pre></div></div>
<p>The default priors are <code>normal(0,10)</code> for the intercept and <code>normal(0, 2.5)</code> for the slope. On a first glance, those may seem fairly informative, but it’s important to note that <code>rstanarm</code> automatically centers and scales the data for the model variables (this helps with model convergence), in which case these are weakly informative priors which are suitable for our research question here and the data we have on plant species richness.</p>
<p>If you would like to change the priors, you can add code, for example <code>prior = normal(0, 1), prior_intercept = normal(0, 5)</code> inside the <code>stan_glm()</code> function.</p>
<p>When using packages like <code>rstanarm</code> and <code>brms</code> which you will see in a bit, it’s a good idea to actually look at the <code>Stan</code> code behind the model.</p>
<p>You can extract the code with the following function:</p>
<div><div><pre><code><span>stancode</span><span> </span><span>&lt;-</span><span> </span><span>rstan</span><span>::</span><span>get_stancode</span><span>(</span><span>stan_glm1</span><span>$</span><span>stanfit</span><span>)</span><span>
</span><span>cat</span><span>(</span><span>stancode</span><span>)</span><span>
</span></code></pre></div></div>
<p>You’ll see a lot of code appear in the console - this is what <code>rstanarm</code> has “written” for you. Amidst the many letters, you can see that the overall structure is like the <code>Stan</code> models we wrote in <a href="https://ourcodingclub.github.io/2018/04/17/stan-intro.html" target="_blank">our intro <code>Stan</code> tutorial</a> - first, we state the parameters for the data, the data gets transformed (scaled and centered), then we define our model and finally, we calculate the predictions in the generated quantities block.</p>
<h2>Explore a model using a negative binomial distribution</h2>
<p>Up until now, we’ve been using a <code>Poisson</code> distribution for our models, which is suitable for when we are working with discrete count data. The <a href="https://en.wikipedia.org/wiki/Poisson_distribution" target="_blank">Poisson distribution</a> assumes that the mean and variance are the same. In ecology and probably other disciplines too, the data might have variance greater than the mean. In this case, we call the data <a href="https://en.wikipedia.org/wiki/Overdispersion">overdispersed</a>. The negative binomial distribution adjusts the variance independently from the mean and as such is more flexible than Poisson. The Poisson distribution is actually a type of a negative binomial distribution.</p>
<p><strong>In <code>rstanarm</code>, it’s easy to <em>update</em> a model using a different data distribution. In our case, we can try a negative binomial distribution.</strong></p>
<div><div><pre><code><span>stan_glm2</span><span> </span><span>&lt;-</span><span> </span><span>update</span><span>(</span><span>stan_glm1</span><span>,</span><span> </span><span>family</span><span> </span><span>=</span><span> </span><span>neg_binomial_2</span><span>)</span><span>
</span><span># Check convergence &amp; priors</span><span>
</span><span>plot</span><span>(</span><span>stan_glm2</span><span>,</span><span> </span><span>plotfun</span><span> </span><span>=</span><span> </span><span>"trace"</span><span>)</span><span>
</span><span>summary</span><span>(</span><span>stan_glm2</span><span>)</span><span>
</span><span>prior_summary</span><span>(</span><span>stan_glm2</span><span>)</span><span>
</span><span># Posterior Predictive Checks</span><span>
</span><span>pp_check</span><span>(</span><span>stan_glm2</span><span>,</span><span> </span><span>plotfun</span><span> </span><span>=</span><span> </span><span>"stat"</span><span>,</span><span> </span><span>stat</span><span> </span><span>=</span><span> </span><span>"mean"</span><span>)</span><span>
</span><span>pp_check</span><span>(</span><span>stan_glm2</span><span>,</span><span> </span><span>plotfun</span><span> </span><span>=</span><span> </span><span>"dens_overlay"</span><span>)</span><span>
</span></code></pre></div></div>
<center>
<img alt="Img" src="https://ourcodingclub.github.io/img/stan2_hist.png"> <img alt="Img" src="https://ourcodingclub.github.io/img/stan2_density.png"> </center>
<p>You can see that the model outputs are very similar - this is to be expected, because the Poisson distribution is actually a type of a negative binomial distribution. Once again, the model is underpredicting the frequency of low species richness values.</p>
<h2>Run a <code>Stan</code> model using the <code>brms</code> package</h2>
<p><code>brms</code> is another package that serves a similar purpose to <code>rstanarm</code> - it allows you to run <code>Stan</code> models using simple code syntax. <code>brms</code> writes all Stan models from scratch and has to compile them, while <code>rstanarm</code> comes with precompiled code (so when we were running our <code>rstanarm</code> models earlier, you didn’t see any messages about <code>C++</code> compiling, since that was already done in advance).</p>
<p>The results of the models ran using the different packages should not be different any more than is to be expected by chance (every time you re-run a model using Bayesian inference, the results will be a tiny bit different, because of how the posterior sampling is done, but those differences are of very small magnitudes, like between <code>0.19887</code> and <code>0.19884</code>).</p>
<p>One advantage of <code>brms</code> is that the <code>Stan</code> code you can extract from it is much easier to read than the <code>Stan</code> code coming from <code>rstanarm</code>. We can fit a <code>brms</code> model to check it out.</p>
<div><div><pre><code><span>stan_glm_brms</span><span> </span><span>&lt;-</span><span> </span><span>brm</span><span>(</span><span>bf</span><span>(</span><span>Richness</span><span> </span><span>~</span><span> </span><span>I</span><span>(</span><span>Year</span><span>-2007</span><span>),</span><span>
</span><span>family</span><span> </span><span>=</span><span> </span><span>brmsfamily</span><span>(</span><span>'poisson'</span><span>)),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>toolik_richness</span><span>,</span><span>
</span><span>iter</span><span> </span><span>=</span><span> </span><span>1000</span><span>,</span><span>
</span><span>chains</span><span> </span><span>=</span><span> </span><span>4</span><span>,</span><span> </span><span>cores</span><span> </span><span>=</span><span> </span><span>4</span><span>)</span><span>
</span><span>summary</span><span>(</span><span>stan_glm_brms</span><span>)</span><span>
</span><span>plot</span><span>(</span><span>stan_glm_brms</span><span>)</span><span>
</span><span># Extract Stan code</span><span>
</span><span># The code is nicely annotated, so you can read through</span><span>
</span><span>stancode</span><span>(</span><span>stan_glm_brms</span><span>)</span><span>
</span></code></pre></div></div>
<h2>Conclusion</h2>
<p>In this tutorial we learned to fit <code>Stan</code> models in <code>R</code> using the <code>rstanarm</code> and <code>brms</code> packages which write the <code>Stan</code> code for us, so they can be seen as a gentler introduction to <code>Stan</code>. We looked at two different data distributions that are suitable for left-skewed discrete count data - <code>Poisson</code> and <code>negative binomial</code>. If you are keen to get back into <code>Stan</code> coding, we’ve showed how to extract the <code>Stan</code> code behind the <code>rstanarm</code> and <code>brms</code> models, so you can take that code and develop it further. You can check out our other <a href="https://ourcodingclub.github.io/2018/04/17/stan-intro.html" target="_blank"><code>Stan</code> tutorial</a>. to see how to write and run <code>Stan</code> code using the <code>rstan</code> package.</p>
<h2>Further resources</h2>
<p><a href="https://cran.r-project.org/web/packages/rstanarm/rstanarm.pdf" target="_blank">rstanarm vignette</a></p>
<p><a href="https://cran.r-project.org/web/packages/brms/brms.pdf" target="_blank">brms vignette</a></p>
<p><a href="https://cran.r-project.org/web/packages/rstanarm/vignettes/rstanarm.html" target="_blank">How to Use the rstanarm Package by Jonah Gabry and Ben Goodrich </a></p>
<p><a href="http://mc-stan.org/" target="_blank">Stan website</a></p>
<p><a href="https://github.com/stan-dev/stan/releases/download/v2.14.0/stan-reference-2.14.0.pdf" target="_blank">Stan manual(v2.14)</a></p>
<p><a href="https://cran.r-project.org/web/packages/rstan/vignettes/rstan.html" target="_blank">Rstan vignette</a></p>
<p><a href="https://t.co/6d3omvBkrd" target="_blank">STANCON 2017 Intro Course Materials</a></p>
<p><a href="http://xcelab.net/rm/statistical-rethinking/" target="_blank">Statistical Rethinking by R. McElreath</a></p>
<p><a href="https://groups.google.com/forum/#!forum/stan-users" target="_blank">Stan mailing list</a></p>
<hr>
<hr>
<p><strong>Check out <a href="https://ourcodingclub.github.io/workshop/" target="_blank">this page</a> to learn how you can get involved! We are very happy to have people use our tutorials and adapt them to their needs. We are also very keen to expand the content on the website, so feel free to get in touch if you’d like to write a tutorial!</strong></p>
<p>This work is licensed under a <a href="https://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>. <a href="https://creativecommons.org/licenses/by-sa/4.0/"><img alt="Img" src="https://licensebuttons.net/l/by-sa/4.0/80x15.png"></a></p>
<h3><a href="https://www.surveymonkey.co.uk/r/PX3XHD2" target="_blank">  We would love to hear your feedback, please fill out our survey!</a></h3>
<h3>  You can contact us with any questions on <a href="mailto:ourcodingclub@gmail.com?Subject=Tutorial%20question" target="_top">ourcodingclub@gmail.com</a>
</h3>
<h3>  Related tutorials:</h3>
<ul>
<li>
<a href="http://ourcodingclub.github.io/2019/03/26/spatial.html">
  - Intro to spatial analysis in R
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/12/02/tidyverse.html">
  - Advanced data manipulation and visualisation
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/11/26/earth-engine.html">
  - Intro to the Google Earth Engine
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/10/15/machine-learning.html">
  - Intro to Machine Learning in R (K Nearest Neighbours Algorithm)
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/04/30/stan-2.html">
  - Generalised linear models in Stan
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/04/17/stan-intro.html">
  - Intro to Stan
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/04/06/model-design.html">
  - Intro to model design
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/03/06/tidyverse.html">
  - GitHub, Tidyverse and Markdown - a coding workshop for the EVENET network
</a>
</li>
</ul>
<h3>  Subscribe to our mailing list:</h3>
</div>
</div>
</div>
</div>
</section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via ourcodingclub.github.io http://bit.ly/2uAMVVw<br>
<br>
May 5, 2019 at 08:19AM
